lib/std/debug.zig

```diff
@@ -50,163 +40,378 @@ pub const LineInfo = struct {
     }
 };
 
-var stderr_mutex = std.Thread.Mutex{};
+pub const SymbolInfo = struct {
+    symbol_name: []const u8 = "???",
+    compile_unit_name: []const u8 = "???",
+    line_info: ?LineInfo = null,
+
+    fn deinit(self: @This()) void {
+        if (self.line_info) |li| {
+            li.deinit();
+        }
+    }
+};
+
+pub const default_config = struct {
+    pub fn getWriter() std.fs.File.Writer {
+        return std.io.getStdErr().writer();
+    }
+
+    pub fn detectTTYConfig() TTY.Config {
```

Comment:
This is more an implementation detail than part of the debug API, on a freestanding target you may not care about TTYs or colors or escape sequences.

Can we demote this?

Reply:
By demote do you mean removing `pub`? I haven't seen this usage of the word before.

Reply:
Yes, removing it from the public interface.
The TTY details are only useful if we're going to print the trace on a terminal, the default implementation in `debug.zig` will make use of this.

Reply:
I am a bit reluctant to do so for two reasons:
1. This would potentially useful for std.debug.print (though maybe this isn't the best way to access this information)
2. Right now there is also `std.log.detectTTYConfig` which is needed for outputting stack traces to the log. This is used in `GeneralPurposeAllocator` for instance.

Reply:
> This would potentially useful for std.debug.print 

Not really. `debug.print` is a quick way of printing out stuff to stderr (or something equivalent), whether the `stderr` handle is a TTY or not doesn't really matter as it doesn't do much beside relaying the user-supplied data to a writer.

If the function performed something fancy like stripping (or translating) the escape sequences when the output is not a TTY then it would have made sense to keep it.

> Right now there is also std.log.detectTTYConfig which is needed for outputting stack traces to the log.

This is a different but similar can of worms.
The default implementation of `log` outputs to `stderr` so detecting whether colours can be used or not is useful (note, what if we also coloured the log level strings?), but I'd expect the user-supplied to take care of this detail within the `log` function itself.

Reply:
First of all, I now agree that `std.debug.detectTTYConfig` shouldn't be public for the reasons that you have given. I will change that at some point. I think I was forgetting that `std.debug.print` is just for quick and dirty debug (for some reason).

I do still think that `std.log.detectTTYConfig` should be public (and I think you agree).

> The default implementation of `log` outputs to `stderr` so detecting whether colours can be used or not is useful (note, what if we also coloured the log level strings?), but I'd expect the user-supplied to take care of this detail within the `log` function itself.

I need to be clear: this PR adds the ability to directly override the writer used by `log` (as opposed to changing the function itself). I thought I would go ahead and make this mostly unrelated change because I was making a similar change to `std.debug`.

I don't know what you mean when you say "I'd expect the user-supplied...". Do you mean that it should strip out escape codes if needed? That approach seems pretty horrible to me.

Reply:
> I need to be clear: this PR adds the ability to directly override the writer used by log

That's IMO unnecessary, the `log` interface is good as is.

> I do still think that std.log.detectTTYConfig should be public (and I think you agree).

Handling terminal API oddities is not something that should be handled in `debug` nor `log`.
A proper solution would be to encapsulate all this stuff in a `TerminalWriter` where you can choose text colour and stuff, but that's outside the scope of this PR.
The TTY stuff you see in `debug` is needed because the code assumes the output is stderr and we want to make the traces look nice, let's stick to this approach for the moment.

Reply:
So what should the approach be to logging stack traces? This is done in `GeneralPurposeAllocator` for instance. I haven't gotten around to looking for other usages and fixing them. I see three options: 

1. don't use color
2. allow for user specification on a per instance basis (include the TTY.Config as a member of `GeneralPurposeAllocator` so it can be set for example)
3.  keep something like `std.log.detectTTYConfig`

Previously, this code (in `GeneralPurposeAllocator`) implicitly used the `TTY.config` which was detected for `stderr` via the `format` function of `StackTrace` which is obviously wrong when `log` is specified not write to stderr.

Based on what you said, I am gravitating toward removing `std.log.detectTTYConfig` and it adding as a member of `GeneralPurposeAllocator` which defaults to no color (and other similar structs).  A different PR can address generalized terminal color writing for `log`.

> That's IMO unnecessary, the log interface is good as is.

I will move that change to a separate PR because it is now longer relevant for stack trace dumping (if we go with the above approach). I do think that specifying the writer as opposed to the log function makes more sense for many usecases and I think the added complexity of an extra function is worth it*. But that's a discussion for a separate PR.

* This is especially true if the log function itself gets more functionality like begin able to specifiy different log levels for different scopes; this is a feature I have been thinking about.

Reply:
> Based on what you said, I am gravitating toward removing std.log.detectTTYConfig and it adding as a member of GeneralPurposeAllocator which defaults to no color (and other similar structs). A different PR can address generalized terminal color writing for log.

Let's try to keep the focus on debug stuff, `GeneralPurposeAllocator` can simply use the default `StackTrace` formatter for the time being and print colorless traces.
It's a big improvement over the status quo where the trace is formatted into a generic log function but `std.debug.detectTTYConfig` queries the stderr handle...

Reply:
Done.